### ISSUE TITLE
Leaking sockets with hibernate-hikaricp and MariaDB Connector/J using UNIX DOMAIN sockets

### ISSUE THREAD

**Issue by ragoley** [brettwooldridge/HikariCP#957]
[ticket.zip](https://github.com/brettwooldridge/HikariCP/files/1246977/ticket.zip)

#### Environment
```
HikariCP version: 2.3.9
JDK version     : Zulu-7 7.20.0.3
Database        : MariaDB 10.1.26
Driver version  : 1.6.3

<dependency>
<groupId>org.hibernate.javax.persistence</groupId>
<artifactId>hibernate-jpa-2.1-api</artifactId>
<version>1.0.0.Final</version>
</dependency>

<dependency>
<groupId>org.hibernate</groupId>
<artifactId>hibernate-entitymanager</artifactId>
<version>5.1.4.Final</version>
</dependency>

<dependency>
<groupId>org.hibernate</groupId>
<artifactId>hibernate-search-orm</artifactId>
<version>5.5.6.Final</version>
</dependency

<dependency>
<groupId>org.hibernate</groupId>
<artifactId>hibernate-hikaricp</artifactId>
<version>5.1.3.Final</version>
</dependency>

<dependency>
<groupId>org.mariadb.jdbc</groupId>
<artifactId>mariadb-java-client</artifactId>
<version>1.6.3</version>
</dependency

# Used by Connector/J to access Unix Domain sockets
<dependency>
<groupId>net.java.dev.jna</groupId>
<artifactId>jna</artifactId>
<version>4.2.2</version>
</dependency>
<dependency>
<groupId>net.java.dev.jna</groupId>
<artifactId>jna-platform</artifactId>
<version>4.2.2</version>
</dependency



```
-----------------------------------------------------------------------------------------

We have a servlet application running in Tomcat 7 that is leaking socket files.  I have tracked it back to the idle timeout reaping of connections.  The leaked socket fd count increases by exactly the number of connections in the hibernate.hikari.maximumPoolSize setting (20) at exactly the idle timeout reaping interval.  This happens without the application logged in or doing anything.  You can watch the /proc/$PID/fd directory and just see the file descriptor count increase.  I will attach pom and other setup files for reference.

I am unsure if this is completely from the JDBC driver or from how the jdbc driver is used in hikaricp.  This only happens when using the MariaDB connector/j driver with a url that uses unix domain sockets.  It does not happen when using TCP/IP connections to localhost.

**Comment by brettwooldridge** [comment]
@ragoley This is almost certainly out of the realm of HikariCP's control.  HikariCP calls ``Connection.close()`` when connections are retired, guaranteed.  You can enable DEBUG level logging for the package ``com.zaxxer.hikari`` and watch.  Debug logging in HikariCP is intentionally minimal, so your logs won't blow up if you turn it on, even in production.

Lastly, v2.3.9 is *very* old.  As you seem to be on a Java 7 compatible JVM, I *strongly* recommend upgrading to v2.4.12.  There is always a small possibility that there was a bug in v2.3.9.

**Comment by ragoley** [comment]
@brettwooldridge I can confirm that switching to 2.4.13 resolves this problem.  It was the only change I made that changed the results any.  For some reason, the 2.3.x hikaricp included from hibernate-hikaricp 5.1.3 when used with the MariaDB connector J driver (tested thru 1.6.5) creates/leaves new socket file descriptors when using the JNA based unix domain socket options.  All the existing file descriptors stay open and it opens exactly the minimumIdle number of new socket file descriptors every time the maxLifetime period is reached.  That does not happen under 2.4.13.  Using 2.4.x solves this issue for my purposes.  Wanted to give you an update in case something else was reported related to 2.3.x.
